### PR TITLE
Improve routes management

### LINKED ISSUE #90: Add support for multiple host per-service
## Purpose

We currently only support one host per service definition to create underlying routes. As some apps (like `edxapp`) may have many routes pointing to the same service (_e.g._ nginx), we need to improve hosts management.

## Proposal

Move the `app.service.host` type from a simple string to a list of dictionaries with the route, the targeted service name and the targeted port.

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

Listing routes from the `apps` global variable definition was not that explicit nor easy to maintain. Hence, following our global "app-driven" logic, we decided to manage routes from their definition stated in an application service template. 

## Proposal

- [x] upgrade routes discovery and creation
- [x] remove hosts/ports definitions in the `apps` global var
- [x] add routes for every application service that requires one or more

As a consequence, this PR will fix #90.

### CODE REVIEW

File: apps/richie/templates/nginx/route.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,43 @@
+{%- set host = richie_host -%}
+
+{#
+    Host prefix can be: previous, next or current. You can use this template with
+    either a prefix set using the following variables (in precedence order):
+
+        - item.1 (in a nested loop)
```
Comment (sampaccoud) [inline apps/richie/templates/nginx/route.yml.j2]: I don't understand "in a nested loop"...

Comment (sampaccoud) [reply apps/richie/templates/nginx/route.yml.j2]: can item be a namedtuple of a dictionary so we write something else than "item.1" !? Can item have a better name also ? it's an item of what?

Comment (jmaupetit) [reply apps/richie/templates/nginx/route.yml.j2]: I'm referring to Ansible nested loops: https://docs.ansible.com/ansible/2.5/user_guide/playbooks_loops.html#complex-loops

But I got your point and I'll come with a better idea / implementation.

File: group_vars/customer/hello/main.yml [hunk 2]
```diff
@@ -6,14 +6,6 @@ project_description: "A minimal project with a hello-world app ({{ env_type }})"
...
-    services:
-      - name:       hello
-        host:       "hello.{{ project_name }}.{{ domain_name }}"
-
-## _____ soon deprecated below _________________________________________________
-hello_host: "{{ project_name }}.{{ domain_name }}"
-hello_port: "8080"
-hello_msg:  "Hello OpenShift! by Arnold"
 
 nginx_ports:
```
Comment (cchaudier) [inline group_vars/customer/hello/main.yml:10]: This also  soon deprecated

File: group_vars/all/main.yml [hunk 3]
```diff
@@ -34,6 +34,12 @@ endpoint_mongodb_ip:     192.168.10.11
 endpoint_mysql_ip:       192.168.10.254
 endpoint_postgresql_ip:  192.168.0.1
 
+# Blue/Green deployment route prefixes
+blue_green_route_prefixes:
+  - "previous"
+  - "current"
+  - "next"
```
Comment (cchaudier) [inline group_vars/all/main.yml:41]: I'v do this before but 
@sampaccoud on 12 Jun  ask this :
I'm skeptical that this should be configurable. Keep it simple and impose it?

See : https://github.com/openfun/arnold/pull/60

File: plugins/filter/deploy.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+
+# pylint: disable=no-self-use,too-few-public-methods
+class FilterModule(object):
+    """Filters used for deployments"""
+
+    def filters(self):
+        """List plugin filters"""
+
+        return {"blue_green_host": blue_green_host}
```
Comment (cchaudier) [inline plugins/filter/deploy.py:46]: This  filter is just for not using the nested loop ?

File: apps/edxapp/templates/nginx/route_cms.yml.j2 [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    app: "edxapp"
+    service: "nginx"
+    version: "{{ edxapp_nginx_image_tag }}"
+    deployment_stamp: "{{ deployment_stamp }}"
+    route_prefix: "{{ prefix }}"
+    route_target_service: "cms"
+  annotations:
+    kubernetes.io/tls-acme: "true"
+spec:
+  host: "{{ edxapp_cms_host | blue_green_host(prefix) }}"
```
Comment (cchaudier) [inline apps/edxapp/templates/nginx/route_cms.yml.j2:18]: This this the host can't be overrided in the apps dict.
If whe want otherride this host we must declare `edxapp_cms_host` in the group_vars.

**Review by sampaccoud — APPROVED** [review]

**Review by cchaudier — COMMENTED** [review]
Huge job ! :+1: 
I've somme comments

File: apps/edxapp/templates/nginx/route_cms.yml.j2 (hunk 5 above)
Comment (jmaupetit) [reply apps/edxapp/templates/nginx/route_cms.yml.j2:18]: >  If whe want otherride this host we must declare edxapp_cms_host in the group_vars.

Yes.

File: plugins/filter/deploy.py (hunk 4 above)
Comment (jmaupetit) [reply plugins/filter/deploy.py:46]: No this is to avoid adding logic to the template.

File: group_vars/all/main.yml (hunk 3 above)
Comment (jmaupetit) [reply group_vars/all/main.yml:41]: :thinking: it might evolve...

Comment (cchaudier) [reply group_vars/all/main.yml:41]: Probably for me it's great like this ^^

File: plugins/filter/deploy.py (hunk 4 above)
Comment (cchaudier) [reply plugins/filter/deploy.py:46]: Ok in this case

File: apps/edxapp/templates/nginx/route_cms.yml.j2 (hunk 5 above)
Comment (cchaudier) [reply apps/edxapp/templates/nginx/route_cms.yml.j2:18]: Ok but I think the apps dict lose is first interest after your PR.

Comment (jmaupetit) [reply apps/edxapp/templates/nginx/route_cms.yml.j2:18]: I disagree and you will see why in the next PR.

**Pull request merged**